packages/lit-html/src/test/helpers_test.ts

```diff
@@ -0,0 +1,151 @@
+  };
+
+  setup(() => {
+    container = document.createElement('div');
+  });
+
+  suite('when', () => {
+    test('true case, without false case', () => {
+      assertRender(
+        when(true, () => html`X`),
+        'X'
+      );
+    });
+
+    test('true case, with false case', () => {
+      assertRender(
+        when(
+          true,
+          () => html`X`,
+          () => html`Y`
+        ),
+        'X'
+      );
+    });
+
+    test('false case', () => {
+      assertRender(
+        when(
+          false,
+          () => html`X`,
+          () => html`Y`
+        ),
+        'Y'
+      );
+    });
+  });
+
+  suite('map', () => {
+    test('with array', () => {
+      assertRender(
+        map(['a', 'b', 'c'], (v) => html`<p>${v}</p>`),
+        '<p>a</p><p>b</p><p>c</p>'
+      );
+    });
+
+    test('with empty array', () => {
+      assertRender(
+        map([], (v) => html`<p>${v}</p>`),
+        ''
+      );
+    });
+
+    test('with undefined', () => {
+      assertRender(
+        map(undefined, (v) => html`<p>${v}</p>`),
+        ''
+      );
+    });
+
+    test('with iterable', () => {
+      function* iterate<T>(items: Array<T>) {
+        for (const i of items) {
+          yield i;
+        }
+      }
+      assertRender(
+        map(iterate(['a', 'b', 'c']), (v) => html`<p>${v}</p>`),
+        '<p>a</p><p>b</p><p>c</p>'
+      );
+    });
+
+    test('passes index', () => {
+      assertRender(
+        map(['a', 'b', 'c'], (v, i) => html`<p>${v}:${i}</p>`),
+        '<p>a:0</p><p>b:1</p><p>c:2</p>'
+      );
+    });
+  });
+
+  suite('join', () => {
+    test('with array', () => {
+      assertRender(join(['a', 'b', 'c'], ','), 'a,b,c');
+    });
+
+    test('with empty array', () => {
+      assertRender(join([], ','), '');
+    });
+
+    test('with undefined', () => {
+      assertRender(join(undefined, ','), '');
+    });
+
+    test('with iterable', () => {
+      function* iterate<T>(items: Array<T>) {
+        for (const i of items) {
+          yield i;
+        }
+      }
+      assertRender(join(iterate(['a', 'b', 'c']), ','), 'a,b,c');
+    });
+
+    test('passes index', () => {
+      assertRender(
+        join(['a', 'b', 'c'], (i) => html`<p>${i}</p>`),
+        'a<p>0</p>b<p>1</p>c'
+      );
+    });
+  });
+
+  suite('range', () => {
+    test('positive end', () => {
+      assert.deepEqual([...range(0)], []);
+      assert.deepEqual([...range(3)], [0, 1, 2]);
+    });
+
+    test('start and end', () => {
+      assert.deepEqual([...range(0, 3)], [0, 1, 2]);
+      assert.deepEqual([...range(-1, 1)], [-1, 0]);
+      assert.deepEqual([...range(-2, -1)], [-2]);
+    });
+
+    test('end < start', () => {
+      assert.deepEqual([...range(2, 1)], []);
+    });
+
+    test('custom step', () => {
+      assert.deepEqual([...range(0, 10, 3)], [0, 3, 6, 9]);
+    });
+
+    test('negative step', () => {
+      assert.deepEqual([...range(0, -3, -1)], [0, -1, -2]);
```

Comment:
Yikes, does e.g. `range(0, 10, -1)` this need any infinite-loop protection? At least in DEBUG mode?

Reply:
I'd suggest matching the behavior of range here: https://github.com/google/closure-library/blob/master/closure/goog/array/array.js#L1570

Reply:
Added a case to this test. There was already one test for infinitely loop protection in the "end < start" case above.

The reason this already works is because of 1) the `<` test, which means that if `end < start` the loop breaks immediately, and 2) the switch of the operands for the `<` test when `step` is negative, which applies that same test correctly in that case. It's the ternary in here:

```ts
  for (let i = start; step > 0 ? i < end : end < i; i += step) {
  ```

Reply:
👍